File: Task 3/C++Stock Viewer/Product/Product.cpp
#include "Product.h"
Product::Product(){
    cost = 0;
    name = "";
    supplier = "";
    packaging = Packaging(0, 0);
}

Product::Product(std::string name, float cost, std::string supplier, float discount_amount, int packing_units, float cost_per_unit){
    this->cost = cost;
    this->name = name;
    this->supplier = supplier;
    this->packaging = Packaging(packing_units, cost_per_unit);
}

Product::Product(const Product& other) {
    // Copy the values of the other object
    this->cost = other.cost;
    this->name = other.name;
    this->supplier = other.supplier;
    this->discount_amount = other.discount_amount;
    this->packaging = Packaging(other.packaging);

}

float Product::calculate_discount() {
    // Get the current time
    std::time_t t = std::time(nullptr);

    // Convert the current time to a structure
    std::tm* now = std::localtime(&t);

    // Get the current month
    int currentMonth = now->tm_mon;

    // Check if the current month is December (index 11)
    if (currentMonth == 11 && (discount_amount >= 0 && discount_amount <= 1)){
       return (1 - discount_amount);;
    }else{
        return 1;
    };
}

std::string Product::display_product_info(){
    return  "Product: "+ name + 
            "\nCost: " +std::to_string(cost)+ 
            "\nSupplier: "+supplier+ 
            "\nDiscount Amount: "+std::to_string(discount_amount)+
            "\n" +packaging.display_product_info();
};

File: Task 1/Java/logistics/src/main/java/com/logistics/Components/Product/Product.java
package com.logistics.Components.Product;

import java.util.Calendar;

public class Product {
    protected float cost;
    protected String name;
    protected String supplier;
    protected float discountAmount;
    protected Packaging packaging;

    public Product() {
        cost = 0;
        name = "";
        supplier = "";
        discountAmount = 1;
        packaging = new Packaging(0, 0);
    }

    public Product(String name, float cost, String supplier, float discountAmount, int packingUnits, float costPerUnit) {
        this.name = name;
        this.cost = cost;
        this.supplier = supplier;
        this.discountAmount = discountAmount;
        this.packaging = new Packaging(packingUnits, costPerUnit);
    }

    public Product(Product other) {
        this.name = other.name;
        this.cost = other.cost;
        this.supplier = other.supplier;
        this.discountAmount = other.discountAmount;
        this.packaging = new Packaging(other.packaging);
    }

    // Getters
    public float getCost() { return cost; }
    public String getName() { return name; }
    public String getSupplier() { return supplier; }
    public float getDiscountAmount(){return discountAmount;}
    public float getPackagingCost() { return packaging.calculateTotalCost(); }

    // Setters
    public void setCost(float cost) { this.cost = cost; }
    public void setName(String name) { this.name = name; }
    public void setSupplier(String supplier) { this.supplier = supplier; }
    public void setDiscountAmount(float discountAmount){this.discountAmount = discountAmount;}
    public void setPackagingData(int packingUnits, float costPerUnit) { this.packaging = new Packaging(packingUnits, costPerUnit); }

    // Clone method
    public Product clone() {
        return new Product(this);
    }

    public float calculateDiscount() {
        // Get the current month
        int currentMonth = Calendar.getInstance().get(Calendar.MONTH);

        // Check if the current month is December 
        if (currentMonth == Calendar.DECEMBER && (discountAmount >= 0 && discountAmount <= 1)) {
            return (1 - discountAmount);
        } else {
            return 1;
        }
    }

    public String display_product_info() {
        return "\nName: " + name +
               "\nCost: " + cost +
               "\nSupplier: " + supplier +
               "\nDiscount Amount: " + calculateDiscount() +
               "\nPackaging Cost: " + getPackagingCost();
    }
}
